### PR TITLE
Feature/pluggable timeliness

### LINKED ISSUE #13: Refactoring "timeliness" as a dimension for assessment
# Description

Timeliness of publication, along with Structure and Schema of the source files, is a dimension used in the assessment of data quality in the first dashboard we've developed for UK25k spend data https://github.com/okfn/data-quality-uk-25k-spend

However, timeliness is really hard to a) detect from the data available for a given data file, and b) to generalise for use over a range of different datasets we'd want to assess for quality.

Using the ideas we developed for the uk25k spend data, we need to address two points:
- Configuration of "timeliness" as a dimension for assessment
- Strategies for detecting "timely publication" for a given file
# Tasks
- [ ] Timeliness configuration (on `dq.json`)
  - [ ] Assess timeliness [ `assess_timeliness: False` ]: boolean. Whether to include timeliness as a dimension of the data quality assessment. Default to false.
  - [ ] Timeliness period [ `timeliness_period`: 1 ]: integer. The period of time in which a file publication can be considered timely. In months. Default 1.
  - [ ] Timeliness strategy [ `timeliness_strategy` ]: 'parse_filename'. The strategy to use to detect the period the data file applies to. One of `parse_filename`, `parse_property` (still TBD in detail)
    - [ ] Implement the parse filename strategy which should be able to be extracted from the uk25k spend codebase
- [ ] Apply the changes here to the uk25k spend data
- [ ] Tests

### PR DESCRIPTION
**Pull request by georgiana-b**
This pull request fixes #13.

Implemented: 
- I've replaced what we currently aggregate on and call `period_id` with `created_at`
- I added support for timeliness configuration options
- I built functionality to extract the `period_id` according to the `timeliness_strategy` 
- I've added tests for the `period_id` extraction
- Abort timeliness assessment if percent of sources lacking `period_id` is greater than the one specified
- Also use the detection mechanism for user-provided `period-id` and overwrite  the provided value with the same dates but different format (to avoid parsing it again in the Aggregator)

### CODE REVIEW

File: data_quality/tasks/extract_relevance_period.py [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+        for source in sources:
+            extracted_period = self.resolve_period(source)
+            if not extracted_period:
+                period = ''
+            else:
+                period_start, period_end = extracted_period
+                period = '/'.join([compat.str(period_start.date()),
+                                   compat.str(period_end.date())])
+            source['period_id'] = period
+        return sources
```
Comment (georgiana-b) [inline data_quality/tasks/extract_relevance_period.py:89]: Not sure what to do here. Normally, we would like to write the updated sources in `source_file` overwriting the existing `period_id`. However, that might be bad for users who have their own `period_id`. I think we could add another configuration option: `extract_period: true/false`. WDYT?

Comment (pwalsh) [inline data_quality/tasks/extract_relevance_period.py:89]: ok.

File: data_quality/tasks/assess_performance.py [hunk 2]
```diff
@@ -17,10 +17,18 @@ class PerformanceAssessor(Task):
 
     """A Task runner to assess and write the performance of publishers for each
        period.
+
+    Note the distiction between a source's creation period
+     (referenced here as "period") and the period the content of a source
+     reffers to (referenced here as "relevance_period").
```
Comment (pwalsh) [inline data_quality/tasks/assess_performance.py]: "refers"

File: data_quality/tasks/assess_performance.py [hunk 3]
```diff
@@ -17,10 +17,18 @@ class PerformanceAssessor(Task):
...
     """
 
     def __init__(self, *args, **kwargs):
         super(PerformanceAssessor, self).__init__(*args, **kwargs)
+        self.assess_timeliness = self.config['assess_timeliness']
+        timeliness_params = self.config['timeliness']
+        self.timeliness_period = timeliness_params.get('timeliness_period', '1')
+        self.max_empty_period = timeliness_params.get('max_empty_period', '10')
```
Comment (pwalsh) [inline data_quality/tasks/assess_performance.py]: ?

Comment (georgiana-b) [inline data_quality/tasks/assess_performance.py]: @pwalsh  Is this a question about why these lines are here or about `max_empty_period` ?

Comment (pwalsh) [inline data_quality/tasks/assess_performance.py]: `max_empty_period`

Comment (georgiana-b) [inline data_quality/tasks/assess_performance.py]: It is the maximum percent of lines that can have empty `relevance_period` but it wasn't clear from the name. I changed it to `max_empty_relevance_period`.

Comment (pwalsh) [inline data_quality/tasks/assess_performance.py]: ok

**Comment by pwalsh** [comment]
@georgiana-b is this ready for review (I refer to the "partially fixes" part of the original message above)?

**Comment by georgiana-b** [comment]
@pwalsh I realized that I misplaced some functionality. I will solve that and let you know when it's ready for review. Apologies for making it confusing.

**Comment by georgiana-b** [comment]
@pwalsh  Ready for review

File: data_quality/tasks/aggregate.py [hunk 4]
```diff
@@ -37,6 +38,8 @@ def __init__(self, config, **kwargs):
         self.run_id = compat.str(uuid.uuid4().hex)
         self.timestamp = datetime.datetime.now(pytz.utc)
         self.max_score = 10
+        self.assess_timeliness = self.config['assess_timeliness']
+        self.timeliness_period = self.config['timeliness'].get('timeliness_period', '1')
```
Comment (pwalsh) [inline data_quality/tasks/aggregate.py]: Why `'1'` and not `1`?

File: data_quality/tasks/aggregate.py [hunk 5]
```diff
@@ -69,7 +76,8 @@ def run(self, pipeline):
 
     def get_lookup(self):
 
-        _keys = ['id', 'publisher_id', self.data_key, 'period_id', 'title']
+        _keys = ['id', 'publisher_id', self.data_key, 'created_at', 'title',
```
Comment (pwalsh) [inline data_quality/tasks/aggregate.py:81]: btw, why is this list of keys not coming from the json table schema?

**Comment by pwalsh** [comment]
@georgiana-b ready to merge, just two small questions.

File: data_quality/tasks/aggregate.py (hunk 5 above)
Comment (georgiana-b) [inline data_quality/tasks/aggregate.py:81]: Because it is a subset specific to this task.

File: data_quality/tasks/aggregate.py (hunk 4 above)
Comment (georgiana-b) [inline data_quality/tasks/aggregate.py]: This was indeed by mistake. I corrected it.

**Comment by pwalsh** [comment]
@georgiana-b this branch has merge conflicts with master. pls fix.

**Pull request merged**